src/wallet/wallet.cpp

```diff
@@ -1004,6 +1037,26 @@ bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockI
         if (fExisted && !fUpdate) return false;
         if (fExisted || IsMine(tx) || IsFromMe(tx))
         {
+            /* Check if any keys in the wallet keypool that were supposed to be unused
+             * have appeared in a new transaction. If so, remove those keys from the keypool.
+             * This can happen when restoring an old wallet backup that does not contain
+             * the mostly recently created transactions from newer versions of the wallet.
+             */
+            std::set<CKeyID> keyPool;
+            GetAllReserveKeys(keyPool);
```

Comment:
Yes, you're right that this would be much more efficient. However, `GetAllReserveKeys()` is also called elsewhere, which would also need to be modified to accept a map, so I'd prefer not to change it as part of this PR.

This can be fixed in a follow-up PR unless you think the performance in `MarkReserveKeysAsUsed()` is unacceptable.